pkg/clustermgmt/clusterWatch.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	// We update by name, and the name *should be* the same for a given cluster
+	// Objects from a given cluster collide and update rather than duplicate insert
+	// Unmarshall ManagedCluster
+
+	var resource model.Resource
+	managedCluster := clusterv1.ManagedCluster{}
+	err = json.Unmarshal(j, &managedCluster)
+	if err != nil {
+		klog.Warning("Failed to Unmarshal MangedCluster", err)
+	}
+	klog.Info("Received ManagedCluster: ", managedCluster.Name)
+
+	resource = transformManagedCluster(&managedCluster)
+
+	dao := database.NewDAO(nil)
+
+	// Upsert (attempt update, attempt insert on failure)
+	dao.UpsertCluster(resource)
+
+	// If a cluster is offline we remove the resources from that cluster, but leave the cluster resource object.
+	/*if resource.Properties["status"] == "offline" {
+		klog.Infof("Cluster %s is offline, removing cluster resources from datastore.", cluster.GetName())
+		delClusterResources(cluster)
+	}*/
+
+}
+
+func isClusterMissing(err error) bool {
+	if err == nil {
+		return false
+	}
+	return strings.Contains(err.Error(), "could not find the requested resource")
+}
+
+// Transform ManagedCluster object into db.Resource suitable for insert into DB
+func transformManagedCluster(managedCluster *clusterv1.ManagedCluster) model.Resource {
+	// https://github.com/stolostron/api/blob/main/cluster/v1/types.go#L78
+	// We use ManagedCluster as the primary source of information
+	// Properties duplicated between this and ManagedClusterInfo are taken from ManagedCluster
+	props := make(map[string]interface{})
+
+	if managedCluster.GetLabels() != nil {
+		// Unmarshaling labels to map[string]interface{}
+		var labelMap map[string]interface{}
+		clusterLabels, _ := json.Marshal(managedCluster.GetLabels())
+		err := json.Unmarshal(clusterLabels, &labelMap)
+		if err == nil {
+			props["label"] = labelMap
+		}
+	}
+
+	props["kind"] = "Cluster"
+	props["name"] = managedCluster.GetName()                  // must match ManagedClusterInfo
+	props["_clusterNamespace"] = managedCluster.GetName()     // maps to the namespace of ManagedClusterInfo
+	props["apigroup"] = "internal.open-cluster-management.io" // maps rbac to ManagedClusterInfo
+	props["created"] = managedCluster.GetCreationTimestamp().UTC().Format(time.RFC3339)
+
+	cpuCapacity := managedCluster.Status.Capacity["cpu"]
+	props["cpu"], _ = cpuCapacity.AsInt64()
+	memCapacity := managedCluster.Status.Capacity["memory"]
+	props["memory"] = memCapacity.String()
+	props["kubernetesVersion"] = managedCluster.Status.Version.Kubernetes
+
+	for _, condition := range managedCluster.Status.Conditions {
+		props[condition.Type] = string(condition.Status)
+	}
+	// Fetch corresponding ManagedClusterInfo
+	managedClusterInfoGvr, _ := schema.ParseResourceArg("managedclusterinfos.v1beta1.internal.open-cluster-management.io")
+
+	minfo, err := dynamicClient.Resource(*managedClusterInfoGvr).Namespace(managedCluster.GetName()).Get(context.TODO(), managedCluster.GetName(), v1.GetOptions{})
```

Comment:
This is only triggered when the ManagedCluster resource changes, so if only ManagedClusterInfo changes we won't get the event and we'll be out of sync.
I think the only way to stay in sync is by watching both resources and merging as we get the data.

Reply:
Done